### DISCUSSION TITLE
baseAtom.onMount not called if atom is only used by callback action

### DISCUSSION THREAD

**Discussion by Cpt-Falcon** [pmndrs/jotai#2243] [Bug report] [answered]
Hi,

So I am using an atomWithStorage, and I'm currently seeing an issue with this:

```
export interface IGuidsList {
    configsGuids: string[];
}

export const stateGuids = atomWithStorage<IGuidsList>('stateGuidsAtom', {
    configsGuids: [],
});
```
My atom is only being called through callback actions like this:

```
    const results = useAtomCallback(
        useCallback((get, set) => {
            const guidsList = get(stateGuids);
            return guidsList.configsGuids.map((guid) => ({
                guid: guid,
                label: get(stateIdentifiers(genGuidObject(guid))),
            }));
        }, []),
        {
            store: store,
        },
    );
```

I copied the code for atomWithStorage from this repo as a test. I put a breakpoint and console log in this code.

```
    baseAtom.onMount = (setAtom) => {
        console.log("Atom mounted");
        if (!getOnInit) {
            setAtom(storage.getItem(key, initialValue) as Value | Promise<Value>);
        }
        let unsub: Unsubscribe | undefined;
        if (storage.subscribe) {
            unsub = storage.subscribe(key, setAtom, initialValue);
        }
        return unsub;
    };
```

Its never getting triggered if ONLY called by a callback action. That means that when only called from a callback, the atom never loads the content from storage. If I use a standard useAtomValue in some component, onMount gets called. Shouldn't onMount get called even if only being called from a callback?

I will continue to investigate why this is happening by bringing in more temporary code from the jotai repo.

**Comment by Cpt-Falcon** [comment]
Ok I see the problem. On mount isn't associated with the atomWithStorage at all when being called purely through a callback. The atom must be being cached by the callback somehow before the onMount function has been assigned. That shouldn't be possible, hoping someone may have a possible explanation. See the screenshot below, when read is being called onMount doesn't exist for atomWithStorage

![image](https://github.com/pmndrs/jotai/assets/10540204/b9985eae-874f-4ab7-babf-d4b411b9b56d)

**Comment by Cpt-Falcon** [comment]
I have found a solution, which is to use the unstable init on load option:
```

    {
        unstable_getOnInit: true,
    },
```

I'm not sure about using this in a production application, but this seems to resolve the problem. Still, to me this doesn't resolve the underlying problem even though it addresses the symptom. IMO base on mount should have been called and should exist on the atom regardless of whether its used in a callback vs used in a useAtomValue

**Comment by himself65** [comment]
`onMount` will be called in under `useEffect` (because it should be mounted to some component)

**Reply by Cpt-Falcon** [reply]
@himself65 thanks for your response, but useAtomCallback is an effect hook. Shouldn't it therefore trigger the mount?

**Reply by himself65** [reply]
I think no, useAtomCallback only uses `useCallback` hook, which won't trigger the useEffect

**Reply by himself65** [reply, marked as answer]
Solution is to manually call `useAtomValue(stateGuids)` on top of your code

**Reply by Cpt-Falcon** [reply]
@himself65 So the recommended approach would be to have some logic in App.tsx that just calls  useAtomValue on all atoms to ensure that they've been mounted before they're used in the rest of the application? That way you can guarantee throughout the application that all atoms have been mounted. 

so something like this:

```
export const useAtomInitialization = () => {
       useAtomValue(atom1);
       useAtomValue(atom2);
       useAtomValue(atom3);
       useAtomValue(atom....);
       useAtomValue(atomN);
}
```
